projects/packages/analyzer/src/api/class-plugin-downloader.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Automattic\Jetpack\Analyzer;
+
+class PluginDownloader {
+	private $type;
+
+	public function __construct($type) {
+		$this->type = $type;
+		$this->output_dir = realpath( dirname( __DIR__ ) . '/../output' ) . '/';
+	}
+
+	public function get_version($version) {
+		echo 'Getting plugin ' . $this->type . '::' . $version . "\n";
+
+		$target_path = $this->get_target_path($version);
+		if (file_exists($target_path)) {
+			if( $version !== 'master' ) {
+				return $target_path;
+			}
+
+			echo "Removing " . $target_path . "\n";
+			system( "rm -rf " . escapeshellarg( $target_path ) );
+		}
+
+		$url = $this->build_url($version);
+		$file_path = $this->download_raw($url);
+		$target_path = $this->unzip($file_path, $version);
+		echo 'Done with plugin ' . $this->type . '::' . $version . "\n";
+		return $target_path;
+	}
+
+	private function get_target_path($version) {
+		$target_folder = $this->type . '__' . $version;
+		return $this->output_dir . $target_folder;
+	}
+
+	private function download_raw($url) {
+		$url_filename = basename( parse_url( $url, PHP_URL_PATH ) );
+
+		$file_path = $this->output_dir . $url_filename;
+		if (false === strpos($url_filename, '.zip')) {
+			$file_path = $file_path  . '.zip';
+		}
+		$context = stream_context_create(array('http' => array(
+			'header' => 'User-Agent: jp-analyzer',
+		)));
+		$out = file_get_contents( $url, false, $context );
+		file_put_contents($file_path, $out);
+
+		return $file_path;
+	}
+
+	private function unzip($file_path, $version) {
+		$target_path = $this->get_target_path($version);
+		$zip = new \ZipArchive;
+
+		if ($zip->open($file_path) !== true) {
+			throw new \Exception("OH, NO!");
+		}
+		for($i = 0; $i < $zip->numFiles; $i++) {
+			$filename = $target_path . '/' . str_replace($zip->getNameIndex(0), '', $zip->getNameIndex($i));
```

Comment:
While it seems unlikely based on the current possible zip files, should we care about the possibility that `$zip->getNameIndex(0)` occurs inside the path other than at the start? Like if someday we wind up with a file like "jetpack_vendor/automattic/jetpack-dev/file.php" in the Jetpack zip.

We already do have some cases that would fail if the zip's root directory were "jetpack/" instead of "jetpack-dev/".

Reply:
Sorry, but I'm not sure I'm following.

`unzip()` implementation is based on the assumption that the passed zip file is in fact a zipped directory, which I assume is a standard for WP plugins.

My understanding is that files listed in a `$zip` are always recursive, meaning at the top there would be the least nested file, while at the bottom - the most nested.

So based on the above, I assume `$zip->getNameIndex(0)` will always point to the top directory.

Reply:
Sure. But try [this zip file](https://github.com/Automattic/jetpack/files/7872530/jetpack-dev.zip) with the code you have here. It won't put `file.php` in the right place.

Reply:
🤦 
I thought you meant the order in which files are indexed in `$zip` object. 
Now I see that `str_replace` replaces all occurrences, which could lead to issues if the root directory appears further in the path. 
I switched to `substr_replace` which replaces only the first occurrence.